log/audit.go

```diff
@@ -18,43 +18,712 @@
+}
+
 //msgp:tag json
 //go:generate msgp -d clearomitted -d "timezone utc" $GOFILE
 
+// AuditCategory represents the category of audit event
+//
+//msgp:shim AuditCategory as:string
+type AuditCategory string
+
+const (
+	AuditCategoryConfig         AuditCategory = "config"
+	AuditCategoryUser           AuditCategory = "user"
+	AuditCategoryServiceAccount AuditCategory = "service-account"
+	AuditCategoryPolicy         AuditCategory = "policy"
+	AuditCategoryGroup          AuditCategory = "group"
+	AuditCategoryBucket         AuditCategory = "bucket"
+	AuditCategoryLifecycle      AuditCategory = "lifecycle"
+	AuditCategoryReplication    AuditCategory = "replication"
+	AuditCategoryNotification   AuditCategory = "notification"
+	AuditCategoryEncryption     AuditCategory = "encryption"
+	AuditCategoryCORS           AuditCategory = "cors"
+	AuditCategoryVersioning     AuditCategory = "versioning"
+	AuditCategoryService        AuditCategory = "service"
+	AuditCategoryKMS            AuditCategory = "kms"
+	AuditCategorySiteRepl       AuditCategory = "site-replication"
+	AuditCategoryPool           AuditCategory = "pool"
+	AuditCategoryIDP            AuditCategory = "idp"
+	AuditCategoryLogRecorder    AuditCategory = "log-recorder"
+	AuditCategoryHeal           AuditCategory = "heal"
+)
+
+// AuditAction represents the type of action performed
+//
+//msgp:shim AuditAction as:string
+type AuditAction string
+
+const (
+	AuditActionCreate  AuditAction = "create"
+	AuditActionUpdate  AuditAction = "update"
+	AuditActionDelete  AuditAction = "delete"
+	AuditActionEnable  AuditAction = "enable"
+	AuditActionDisable AuditAction = "disable"
+	AuditActionSet     AuditAction = "set"
+	AuditActionReset   AuditAction = "reset"
+	AuditActionRestore AuditAction = "restore"
+	AuditActionClear   AuditAction = "clear"
+	AuditActionStart   AuditAction = "start"
+	AuditActionStop    AuditAction = "stop"
+	AuditActionRestart AuditAction = "restart"
+	AuditActionAttach  AuditAction = "attach"
+	AuditActionDetach  AuditAction = "detach"
+)
+
 // Audit represents the user triggered audit events
 type Audit struct {
 	Version    string                 `json:"version"`
 	Time       time.Time              `json:"time"`
 	Node       string                 `json:"node,omitempty"`
 	APIName    string                 `json:"apiName,omitempty"`
+	Category   AuditCategory          `json:"category,omitempty"`
+	Action     AuditAction            `json:"action,omitempty"`
 	Bucket     string                 `json:"bucket,omitempty"`
 	Tags       map[string]string      `json:"tags,omitempty"`
 	RequestID  string                 `json:"requestID,omitempty"`
 	ReqClaims  map[string]interface{} `json:"requestClaims,omitempty"`
 	SourceHost string                 `json:"sourceHost,omitempty"`
 	AccessKey  string                 `json:"accessKey,omitempty"`
 	ParentUser string                 `json:"parentUser,omitempty"`
+	Details    interface{}            `json:"details,omitempty"`
```

Comment:
I am still not a particular fan of this being an `any` type. We will not be able to maintain the type across any serialization.

This would be a lot easier to manage:
```suggestion
	Details    struct {
		Config         *ConfigAuditDetails          `json:"config,omitempty"`
		User           *UserAuditDetails            `json:"user,omitempty"`
		ServiceAccount *ServiceAccountAuditDetails  `json:"serviceAccount,omitempty"`
		Policy         *PolicyAuditDetails          `json:"policy,omitempty"`
		Group          *GroupAuditDetails           `json:"group,omitempty"`
		BucketConfig   *BucketConfigAuditDetails    `json:"bucketConfig,omitempty"`
		Service        *ServiceAuditDetails         `json:"service,omitempty"`
		KMS            *KMSAuditDetails             `json:"kms,omitempty"`
		Pool           *PoolAuditDetails            `json:"pool,omitempty"`
		SiteRepl       *SiteReplicationAuditDetails `json:"siteRepl,omitempty"`
		IDP            *IDPAuditDetails             `json:"idp,omitempty"`
		Recorder       *RecorderAuditDetails        `json:"recorder,omitempty"`
		Heal           *HealAuditDetails            `json:"heal,omitempty"`
	} `json:"details,omitempty"`
```